dev_setup.sh

```diff
@@ -0,0 +1,155 @@
+#!/bin/sh
+
+# cryptol setup script
+#
+# Supported distribution(s):
+#  * macOS 14 (AArch64)
+#
+# This script installs everything needed to get a working development
+# environment for cryptol. Any new environment requirements should be
+# added to this script.
+#
+# There is some half-baked support for macOS 12 (x86_64), but it hasn't been
+# tested.
+#
+
+set -e
+
+HERE=$(cd `dirname $0`; pwd)
+LOG=$HERE/dev_setup.log
+ROOT=$HERE
+
+# Create a new, empty file for environment variables
+VAR_FILE=$ROOT/env.sh
+truncate -s 0 $VAR_FILE
+echo "# Generated environment variables. Manual changes will get clobbered by dev_setup.sh" >> $VAR_FILE
+
+WHAT4_SOLVERS_SNAPSHOT="snapshot-20240212"
+WHAT4_SOLVERS_URL="https://github.com/GaloisInc/what4-solvers/releases/download/$WHAT4_SOLVERS_SNAPSHOT/"
+WHAT4_SOLVERS_MACOS_12="macos-12-X64-bin.zip"
+WHAT4_SOLVERS_MACOS_14="macos-14-ARM64-bin.zip"
+
+MACOS14="macos14"
+MACOS12="macos12"
+# Make sure we're running on a supported platform
+case $(uname -s) in
+    Darwin)
+        if [ $(uname -m) = "arm64" ]; then
+            CRYPTOL_PLATFORM=$MACOS14
+        # This is how we'd support macOS 12. Since this hasn't been tested yet,
+        # we withhold official support.
+        # This might bork on something running macOS <12, since we're basing
+        # the it on the hardware, not the specific version.
+        elif [ $(uname -m) = "x86_64" ]; then
+            CRYPTOL_PLATFORM=$MACOS12
+            echo "Unsupported platform" 2>&1; exit 1;
+        else
+            echo "Unsupported platform" 2>&1; exit 1;
+        fi;;
+    *) echo "Unsupported platform" 2>&1; exit 1;;
+esac
+
+function notice {
+    echo "[NOTICE] $*"
+}
+
+# Requires: LOG set to log file path.
+function logged {
+    if ! [ -z "$LOG" ]
+    then
+        # This may or may not be the same directory that this script lives in.
+        mkdir -p `dirname $LOG`
+        echo "$@" >>$LOG
+        if ! "$@" >>$LOG 2>&1
+        then
+            echo
+            echo "An error occurred; please see $LOG"
+            echo "Here are the last 50 lines:"
+            tail -n 50 $LOG
```

Comment:
I'd recommend shortening this tail and/or putting something indicating there was an error closer to the bottom of the output.  The concern here is that the user will just see the tail of the log (which may or may not contain error output) and not see the indication of an error because it is much further up (and possibly scrolled off the visible screen area).

Reply:
FWIW in scripts where I've used this (since this is based on my script), there is no error to see that isn't in the log since the console output of the script is exclusively non-error messages. Of course, whether 50 lines of log will be enough to show an error depends a lot on what you're doing, though. That's why (in my original script) the message tell you where to find the log in case the last 50 lines aren't useful enough to understand what happened.

Reply:
I popped an `[ERROR]` underneath the log anyway so that users who don't routinely read any terminal output beyond 3 lines away will know that something went wrong.